#### Importing pandas

In [1]:
import pandas as pd
import numpy as np

#### Importing the data

In [2]:
geo=pd.read_csv(r"C:\Users\nyang\Downloads\cells_geo.csv")
day1=pd.read_csv(r"C:\Users\nyang\Downloads\Telcom_dataset.csv")
day2=pd.read_csv(r"C:\Users\nyang\Downloads\Telcom_dataset2.csv")
day3=pd.read_csv(r"C:\Users\nyang\Downloads\Telcom_dataset3.csv")
geo_description=pd.read_excel(r"C:\Users\nyang\Downloads\cells_geo_description.xlsx")
CDR_description=pd.read_excel(r"C:\Users\nyang\Downloads\CDR_description.xlsx")

#### Exploring the datasets

##### 1. Geo

In [3]:
geo.head()

;VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE
0  0;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
1  1;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
2  2;ADJAME;In Service;ABIDJAN;"""Abidjan_EST";AG...                                              
3  3;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...                                              
4  4;ASSINIE;In Service;INTERIEUR;Grand-EST;"ASSI...

**Separating the columns**

In [4]:
geo=geo[";VILLES;STATUS;LOCALISATION;DECOUPZONE;ZONENAME;LONGITUDE;LATITUDE;REGION;AREA;CELL_ID;SITE_CODE"].str.split(';',
                                                                                                                      expand=True)

In [5]:
geo.head()

0        1           2          3                4            5   \
0  0   ADJAME  In Service    ABIDJAN  """Abidjan_EST"     AGBANNEW   
1  1   ADJAME  In Service    ABIDJAN  """Abidjan_EST"     AGBANNEW   
2  2   ADJAME  In Service    ABIDJAN  """Abidjan_EST"     AGBANNEW   
3  3  ASSINIE  In Service  INTERIEUR        Grand-EST  "ASSINIE"""   
4  4  ASSINIE  In Service  INTERIEUR        Grand-EST  "ASSINIE"""   

                  6                 7          8  9           10          11  
0  -4.00066666666667  5.36819444444444     LAGUNE     fbb2e75af3  0bb492ef17  
1  -4.00066666666667  5.36819444444444     LAGUNE     1502501d78  0bb492ef17  
2  -4.00066666666667  5.36819444444444     LAGUNE     0c2e182a0c  0bb492ef17  
3  -3.48846666666667  5.15876111111111  SUD COMOE     69bdfc69fd  8a1e6ce31f  
4  -3.48846666666667  5.15876111111111  SUD COMOE     7a1029d2ab  8a1e6ce31f

**Renaming the columns correctly**

In [6]:
geo=geo.rename(columns={1:"VILLES",
                    2:"STATUS",
                    3:"LOCALISATION",
                    4:"DECOUPZONE",
                    5:"ZONENAME",
                    6:"LONGITUDE",
                    7:"LATITUDE",
                    8:"REGION",
                    9:"AREA",
                    10:"CELL_ID",
                    11:"SITE_CODE"
                  })

In [7]:
geo.drop(0,inplace=True, axis=1)

In [8]:
geo.head()

VILLES      STATUS LOCALISATION       DECOUPZONE     ZONENAME  \
0   ADJAME  In Service      ABIDJAN  """Abidjan_EST"     AGBANNEW   
1   ADJAME  In Service      ABIDJAN  """Abidjan_EST"     AGBANNEW   
2   ADJAME  In Service      ABIDJAN  """Abidjan_EST"     AGBANNEW   
3  ASSINIE  In Service    INTERIEUR        Grand-EST  "ASSINIE"""   
4  ASSINIE  In Service    INTERIEUR        Grand-EST  "ASSINIE"""   

           LONGITUDE          LATITUDE     REGION AREA     CELL_ID   SITE_CODE  
0  -4.00066666666667  5.36819444444444     LAGUNE       fbb2e75af3  0bb492ef17  
1  -4.00066666666667  5.36819444444444     LAGUNE       1502501d78  0bb492ef17  
2  -4.00066666666667  5.36819444444444     LAGUNE       0c2e182a0c  0bb492ef17  
3  -3.48846666666667  5.15876111111111  SUD COMOE       69bdfc69fd  8a1e6ce31f  
4  -3.48846666666667  5.15876111111111  SUD COMOE       7a1029d2ab  8a1e6ce31f

**Summary of the Geo data**

In [9]:
geo_description

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [10]:
geo.describe(include="all")

VILLES      STATUS LOCALISATION   DECOUPZONE              ZONENAME  \
count     3974        3974         3974         3974                  3974   
unique     504           2            2           15                    34   
top     COCODY  In Service    INTERIEUR  Grand-OUEST  ZONE D - Grand-OUEST   
freq       354        3907         2305          528                   527   

                LONGITUDE          LATITUDE  REGION           AREA  \
count                3974              3974    3974           3974   
unique               1013              1025      19              7   
top     -6.45038888888889  5.37363888888889  LAGUNE  ABIDJAN NORTH   
freq                   12                12    1850           1222   

           CELL_ID   SITE_CODE  
count         3974        3974  
unique        3916        1029  
top     6750e0fb7b  20387d4292  
freq            39          42

In [11]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   VILLES        3974 non-null   object
 1   STATUS        3974 non-null   object
 2   LOCALISATION  3974 non-null   object
 3   DECOUPZONE    3974 non-null   object
 4   ZONENAME      3974 non-null   object
 5   LONGITUDE     3974 non-null   object
 6   LATITUDE      3974 non-null   object
 7   REGION        3974 non-null   object
 8   AREA          3974 non-null   object
 9   CELL_ID       3974 non-null   object
 10  SITE_CODE     3974 non-null   object
dtypes: object(11)
memory usage: 341.6+ KB


- There are no null values because every row is in string format

In [12]:
geo["DECOUPZONE"].unique()

array(['"""Abidjan_EST"', 'Grand-EST', 'Grand-OUEST', '"""Abidjan_NORD"',
       '"""Abidjan_CENTRE"', 'Abidjan_NORD', 'Abidjan_EST',
       'Abidjan_OUEST', 'Grand-SUD', 'SUD-OUEST', 'Abidjan_CENTRE',
       'Abidjan_SUD', 'Grand-NORD', 'Gd-ABIDJAN', '0'], dtype=object)

In [13]:
geo.replace('"""Abidjan_EST"', "Abidjan_EST",inplace=True)
geo.replace('"""Abidjan_CENTRE"', "Abidjan_CENTRE",inplace=True)
geo.replace('"""Abidjan_NORD"', "Abidjan_NORD",inplace=True)

In [14]:
geo.describe()

VILLES      STATUS LOCALISATION   DECOUPZONE              ZONENAME  \
count     3974        3974         3974         3974                  3974   
unique     504           2            2           12                    34   
top     COCODY  In Service    INTERIEUR  Grand-OUEST  ZONE D - Grand-OUEST   
freq       354        3907         2305          528                   527   

                LONGITUDE          LATITUDE  REGION           AREA  \
count                3974              3974    3974           3974   
unique               1013              1025      19              7   
top     -6.45038888888889  5.37363888888889  LAGUNE  ABIDJAN NORTH   
freq                   12                12    1850           1222   

           CELL_ID   SITE_CODE  
count         3974        3974  
unique        3916        1029  
top     6750e0fb7b  20387d4292  
freq            39          42

In [15]:
geo.replace('"ASSINIE"""', "ASSINIE",inplace=True)
geo.replace('"KRIKOREA"""', "KRIKOREA",inplace=True)

In [16]:
geo["STATUS"].unique()

array(['In Service', ''], dtype=object)

In [17]:
geo["STATUS"].replace("","Not In Service", inplace=True)

In [18]:
geo["STATUS"].unique()

array(['In Service', 'Not In Service'], dtype=object)

In [19]:
geo["LOCALISATION"].unique()

array(['ABIDJAN', 'INTERIEUR'], dtype=object)

In [20]:
geo["REGION"].unique()

array(['LAGUNE', 'SUD COMOE', 'HAUT SASSANDRA', 'MOYEN COMOE', 'AGNEBY',
       'SUD BANDAMA', 'BAS SASSANDRA', 'LACS', 'FROMAGER',
       'VALLEE DU BANDAMA', 'NZI COMOE', 'ZANZAN', 'MOYEN CAVALLY',
       'WORODOUGOU', 'DENGUELE', '18 MONTAGNES', 'SAVANE', 'MARAHOUE',
       'BAFING'], dtype=object)

In [21]:
geo["AREA"].unique()

array(['', 'ABIDJAN NORTH', 'NORTH WEST', 'EAST', 'ABIDJAN SOUTH',
       'SOUTH WEST', 'CENTRE NORTH'], dtype=object)

In [22]:
geo["AREA"].replace('',np.nan,inplace=True)

In [23]:
geo["AREA"].unique()

array([nan, 'ABIDJAN NORTH', 'NORTH WEST', 'EAST', 'ABIDJAN SOUTH',
       'SOUTH WEST', 'CENTRE NORTH'], dtype=object)

In [24]:
geo["LONGITUDE"]=geo["LONGITUDE"].astype(float)
geo["LATITUDE"]=geo["LATITUDE"].astype(float)

In [25]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   VILLES        3974 non-null   object 
 1   STATUS        3974 non-null   object 
 2   LOCALISATION  3974 non-null   object 
 3   DECOUPZONE    3974 non-null   object 
 4   ZONENAME      3974 non-null   object 
 5   LONGITUDE     3974 non-null   float64
 6   LATITUDE      3974 non-null   float64
 7   REGION        3974 non-null   object 
 8   AREA          3951 non-null   object 
 9   CELL_ID       3974 non-null   object 
 10  SITE_CODE     3974 non-null   object 
dtypes: float64(2), object(9)
memory usage: 341.6+ KB


- Note that there are now null values in the Area column

##### 2. Day 1

In [26]:
day1.describe(include="all")

PRODUTC        VALUE               DATETIME CELL_ON_SITE  \
count     5001  5001.000000                   5001         5001   
unique       3          NaN                    708            7   
top      Voice          NaN  2012-05-07 00:00:10.0            1   
freq      2219          NaN                     71         1252   
mean       NaN    25.869626                    NaN          NaN   
std        NaN   101.934502                    NaN          NaN   
min        NaN     0.000000                    NaN          NaN   
25%        NaN     0.000000                    NaN          NaN   
50%        NaN     0.000000                    NaN          NaN   
75%        NaN    25.000000                    NaN          NaN   
max        NaN  4440.000000                    NaN          NaN   

        DW_A_NUMBER_INT  DW_B_NUMBER_INT COUNTRY_A COUNTRY_B     CELL_ID  \
count              5001             5001      5001      5001        5001   
unique             3903             2848         1         2        1790   
top     4e49073a756df8c  5a4dc9e82deb309     19e2e     19e2e  ffa6759bb2   
freq                 63              807      5001      4026         716   
mean                NaN              NaN       NaN       NaN         NaN   
std                 NaN              NaN       NaN       NaN         NaN   
min                 NaN              NaN       NaN       NaN         NaN   
25%                 NaN              NaN       NaN       NaN         NaN   
50%                 NaN              NaN       NaN       NaN         NaN   
75%                 NaN              NaN       NaN       NaN         NaN   
max                 NaN              NaN       NaN       NaN         NaN   

           SITE_ID  
count         4285  
unique        1076  
top     b47a998a3e  
freq            64  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

- There are outliers in the value column. The mean is 25.869626	but the minimum value is 0 and the 75th percentile is 25 yet the maximum value is 4440

In [27]:
day1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUTC          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATETIME         5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELL_ID          5001 non-null   object
 9   SITE_ID          4285 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [28]:
day1.isnull().any()

PRODUTC            False
VALUE              False
DATETIME           False
CELL_ON_SITE       False
DW_A_NUMBER_INT    False
DW_B_NUMBER_INT    False
COUNTRY_A          False
COUNTRY_B          False
CELL_ID            False
SITE_ID             True
dtype: bool

In [29]:
day1.isnull().sum()

PRODUTC              0
VALUE                0
DATETIME             0
CELL_ON_SITE         0
DW_A_NUMBER_INT      0
DW_B_NUMBER_INT      0
COUNTRY_A            0
COUNTRY_B            0
CELL_ID              0
SITE_ID            716
dtype: int64

- There are 716 null values in the site ID column


**Renaming column 1**

In [30]:
day1.rename(columns={"PRODUTC":"PRODUCT"},inplace=True)

In [31]:
day1.head()

PRODUCT  VALUE               DATETIME CELL_ON_SITE  DW_A_NUMBER_INT  \
0   Voice      0  2012-05-06 23:04:37.0            1  49fae048b547780   
1     sms      0  2012-05-06 23:05:25.0            0  7a54a6eaee19e65   
2     sms     25  2012-05-06 23:05:41.0            0  6008887b0207d82   
3   Voice     56  2012-05-06 23:05:42.0            1  7370032e20f4e96   
4   Voice     14  2012-05-06 23:06:29.0            1  966ba4656e38e62   

   DW_B_NUMBER_INT COUNTRY_A COUNTRY_B     CELL_ID     SITE_ID  
0  1a49f8b0ded9de6     19e2e     19e2e  /TJNe+Mmtu  /TJNe+Mmtu  
1  93343ba6e41af95     19e2e     19e2e  /+cKZKIp41  /+cKZKIp41  
2  7aacc9709ce0a20     19e2e     19e2e  /u0FSD+ahi  /u0FSD+ahi  
3  6af4a60e74c91d3     19e2e     19e2e  +cNeJzsTp3  +cNeJzsTp3  
4  02aa736ee555d5c     19e2e     19e2e  051KhYDCpv  051KhYDCpv

#### Changing the datetime column from string to datetime format, and introducing hour column

In [32]:
import datetime

In [33]:
day1["DATETIME"]=pd.to_datetime(day1["DATETIME"]).dt.to_period("d")

In [35]:
day1["DATETIME"]=pd.to_datetime(day1["DATETIME"]).astype(str)

TypeError: Passing PeriodDtype data is invalid. Use `data.to_timestamp()` instead

In [ ]:
day1["DATETIME"]=pd.to_datetime(day1["DATETIME"],dayfirst=True)

In [ ]:
def hr_function(ts):
    return ts.hour

In [ ]:
day1["Hour"]=day1["DATETIME"].apply(hr_function)

##### Day 2

In [ ]:
day2.describe(include="all")

- There are outliers in the value column. The mean is 24.334933	but the minimum value is 0 and the 75th percentile is 25, yet the maximum value is 3380.

- There are null values in the site ID column

In [ ]:
day2["DATETIME"]=pd.to_datetime(day1["DATETIME"]).dt.to_period("d")

In [ ]:
day2["DATETIME"]=pd.to_datetime(day1["DATETIME"]).astype(str)

In [ ]:
day2["DATETIME"]=pd.to_datetime(day1["DATETIME"],dayfirst=True)

In [ ]:
day2["Hour"]=day1["DATETIME"].apply(hr_function)

##### Day 3

In [ ]:
day3.describe(include="all")

- There are outliers in the value column. The mean is 26.441512	but the minimum value is 0 and the 75th percentile is 25, yet the maximum value is 12900.

- There are null values in the site ID column


**Renaming the last column**

In [ ]:
day1.rename(columns={"SIET_ID":"SITE_ID"},inplace=True)

In [ ]:
day3["DATETIME"]=pd.to_datetime(day1["DATETIME"]).dt.to_period("d")

In [ ]:
day3["DATETIME"]=pd.to_datetime(day1["DATETIME"]).astype(str)

In [ ]:
day3["DATETIME"]=pd.to_datetime(day1["DATETIME"],dayfirst=True)

In [ ]:
day3["Hour"]=day1["DATETIME"].apply(hr_function)

#### More exploration

**_Average billing price per day_**

In [ ]:
day1.groupby("PRODUCT")["VALUE"].mean()

In [ ]:
day2.groupby("PRODUCT")["VALUE"].mean()

In [ ]:
day3.groupby("PRODUCT")["VALUE"].mean()

**_Total billing price per day_**

In [ ]:
day1.groupby("PRODUCT")["VALUE"].sum()

In [ ]:
day2.groupby("PRODUCT")["VALUE"].sum()

In [ ]:
day3.groupby("PRODUCT")["VALUE"].sum()

#### Joining the dataframes

In [ ]:
frames=[day1,day2,day3]
all_days=pd.concat(frames,ignore_index=True)